### PR TITLE
Reproduce and fix ByteString initialization issue

### PR DESCRIPTION
**Pull request by raboof**
References https://github.com/akka/akka-grpc/issues/1014, https://github.com/akka/akka-http/issues/3221

(jr: created issue for that case as search engine fodder at #29354)

### CODE REVIEW

**Comment by jrudolph** [comment]
:+1: Tricky to test though, as it will only fail reliably if it is the first test run with that classloader...

**Comment by jrudolph** [comment]
It failed!!!



But only because of scalafmt...

**Comment by raboof** [comment]
I guess we could create a test that creates its own classloader, but I'm not sure it's worth it. With this one you can at least easily reproduce locally.

**Comment by jrudolph** [comment]
I created #29263 as the reference ticket for this issue.

File: akka-actor-tests/src/test/scala/akka/util/ByteStringInitializationSpec.scala [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+package akka.util
+
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.wordspec.AnyWordSpec
+
+class ByteStringInitializationSpec extends AnyWordSpec with Matchers {
+  "ByteString intialization" should {
+    "not get confused by initializing CompactByteString before ByteString" in {
+      ByteString.empty should not be (null)
```
Comment (jrudolph) [inline akka-actor-tests/src/test/scala/akka/util/ByteStringInitializationSpec.scala:13]: Only fails if that line is removed.

**Comment by jrudolph** [comment]
> I guess we could create a test that creates its own classloader, but I'm not sure it's worth it.

Maybe not worth it but still fun!

**Comment by jrudolph** [comment]
> Test PASSed.

Doesn't fail because PR validation only runs on Scala 2.12.

**Review by jrudolph — APPROVED** [review]
LGTM

**Comment by jrudolph** [comment]
Seemed to show up here as well: https://github.com/akka/akka/issues/29347

**Review by ennru — APPROVED** [review]
LGTM.

**Comment by jrudolph** [comment]
Interesting, seems 2.5 didn't have the same bug in the first place:

https://github.com/akka/akka/blob/82533f989e88c320f4a4dbb5694ccd27155a8d26/akka-actor/src/main/scala-2.13%2B/akka/util/ByteString.scala#L749

**Comment by jrudolph** [comment]
It seems 2.13 support was introduced shortly after 2.5 and 2.6 diverged, so the deltas were not quite the same.

**Pull request merged**